Fix LinkPicker freeze when virtual keyboard is hidden

dcalhoun opened the pull request:
## Description
Fixes #37773.

When a devices virtual keyboard is hidden, e.g. when a hardware keyboard is connected, dismissing the `LinkPicker` resulted in the application freezing. The freeze likely originates from an unconsumed `LayoutAnimation` registered during resizing of the `BottomSheet`.

To address this issue, we now avoid registering a `LayoutAnimation` whenever the changes to the header are sub-pixel values, which can result in the `LayoutAnimation` going unconsumed: https://git.io/J9q2G Long-term, we should likely consider refactoring the `BottomSheet` to holistically avoid stacking `LayoutAnimations`: https://git.io/J9q2l


## How has this been tested?
1. Run native block editor with the virtual keyboard hidden.
    1. iOS simulator: Toggle virtual keyboard off: I/O > Keyboard > Toggle Software Keyboard.
    1. iOS physical device: Connect a physical keyboard.<sup>*</sup>
    1. Android physical device: Connect a physical keyboard and disable "Show on-screen keyboard" within the OS keyboard settings.<sup>*</sup>
    1. Android emulator: Disable "Show on-screen keyboard" within the OS keyboard settings.
1. In a Paragraph block, tap on the Link button located in the formatting toolbar.
1. Tap on "Link to" item to open the link picker.
1. Tap on "Cancel" button to dismiss it.
1. Dismiss the block settings by tapping on the backdrop.
1. Verify the app remains interactive, not frozen.

<small>*These specific context were not tested due to lack of necessary hardware.</small>

## Screenshots <!-- if applicable -->
n/a

## Types of changes
Bug fix

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [ ] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/code/native-mobile.md -->

dcalhoun committed:
Fix LinkPicker freeze when virtual keyboard is hidden

When a devices virtual keyboard is hidden, e.g. when a hardware
keyboard is connected, dismissing the `LinkPicker` resulted in the
application freezing. The freeze likely originates from an unconsumed
`LayoutAnimation` registered during resizing of the `BottomSheet`.

To address this issue, we now avoid registering a `LayoutAnimation`
whenever the changes to the header are sub-pixel values, which can
result in the `LayoutAnimation` going unconsumed. https://git.io/J9q2G

Long-term, we should likely consider refactoring the `BottomSheet` to
holistically avoid stacking `LayoutAnimations`: https://git.io/J9q2l

fluiddot commented on packages/components/src/mobile/bottom-sheet/index.native.js:
```diff
@@ -551,6 +554,7 @@ class BottomSheet extends Component {
 					<View
 						style={ styles.header }
 						onLayout={ this.onHeaderLayout }
+						testID="bottom-sheet-header"
```
I'm wondering if we should try to keep test IDs unique to make queries easier to be made in tests. Maybe we could use a suffix like `-header` in case the `testID` prop is passed:
 ```suggestion
						testID={ `${ rest.testID || 'bottom-sheet' }-header` }
```

Wdyt?

dcalhoun replied:
There doesn't appear to be an explicit need for this currently, but I think this change is sensible to include. 👍🏻

fluiddot replied:
@dcalhoun I'm afraid I provided a bad indentation on the suggested change 😭, I'll push a fix to address it asap.

fluiddot replied:
I've just solved it in https://github.com/WordPress/gutenberg/pull/37782/commits/3fdd6b3f64ef8462866636e76a9102389b6877e7, sorry for the mistake.

dcalhoun committed:
Support unique BottomSheet header testID

This allows individual tests to pass a unique, top-level testID for the BottomSheet and have it propagate to the header as well, which may make querying easier.

Co-authored-by: Carlos Garcia <fluiddot@gmail.com>

fluiddot committed:
Fix indentation issue in bottom sheet component

fluiddot reviewed (approved):
LGTM 🎊 !
Thanks @dcalhoun for addressing the issue and introducing a test for the bottom sheet component 🙇 !

I tested the following cases which all of them passed:
- Use an external keyboard on Android device ✅ 
- Disable software keyboard on iOS simulator ✅
- Use an external keyboard on iOS device ✅

And verified that the unit tests passed too ✅ .

fluiddot commented:
@dcalhoun what do you think about creating a GB-mobile PR to include the fix in Gutenberg Mobile?

fluiddot commented:
Additionally, we might consider adding an entry with this fix in the `react-native-editor` changelog file.

dcalhoun commented:
> @dcalhoun what do you think about creating a GB-mobile PR to include the fix in Gutenberg Mobile?

@fluiddot yep, https://github.com/wordpress-mobile/gutenberg-mobile/pull/4443 already exists. 👍🏻 

> Additionally, we might consider adding an entry with this fix in the `react-native-editor` changelog file.

Yes, that makes sense. 👍🏻 Even though this particular bug will never land in a public-facing release of the WordPress apps, I suppose it is worth noting it does technically exist in the `1.69.0` Gutenberg Mobile release.

dcalhoun committed:
Add change log entry

dcalhoun committed:
Add pull request reference to change log entry
